Nav Unification: Bring back environment flag to disable if needed.

obenland opened the pull request:
#### Changes proposed in this Pull Request

* Checks env variable in addition to being a11n.

#### Testing instructions

<!--
Add as many details as possible to help others reproduce the issue and test the fix.
"Before / After" screenshots can also be very helpful when the change is visual.
-->

1. Go to `/home/:site` and select any site
2. Add `?disable-nav-unification` to the URL.
3. Verify that nav unification gets disabled.

Fixes #49123.

Commit:
Bring back environment flag to disable if needed.

obenland commented:
Still needs some testing

tyxla commented on client/layout/index.jsx:
```diff
@@ -164,7 +164,8 @@ class Layout extends Component {
 				this.props.wccomFrom,
-			'is-nav-unification': isAutomatticTeamMember( this.props.teams ),
+			'is-nav-unification':
+				isAutomatticTeamMember( this.props.teams ) && config.isEnabled( 'nav-unification' ),
```
I've noticed there are [other locations](https://github.com/Automattic/wp-calypso/search?q=%22nav-unification%22) where we check for the `nav-unification` feature flag, but we don't account for `isAutomatticTeamMember` there. 

At the same time, this PR enables the flag for all non-production environments. 

Is that the desired behavior? Also, see the next question.

cpapazoglou replied:
Yes this the desired behavior, as currently decided `nav-unification` should be enabled for all a12s. More about why we still use the flag here https://github.com/Automattic/wp-calypso/issues/49123#issuecomment-765353474

tyxla commented on client/layout/index.jsx:
```diff
@@ -164,7 +164,8 @@ class Layout extends Component {
 				this.props.wccomFrom,
-			'is-nav-unification': isAutomatticTeamMember( this.props.teams ),
+			'is-nav-unification':
+				isAutomatticTeamMember( this.props.teams ) && config.isEnabled( 'nav-unification' ),
```
What's the value of `isAutomatticTeamMember` here? Considering that this is only accessible on testing or development environments, and `stage` environments which are accessible only by proxied users (who are a12s anyway), do we really need that check?

cpapazoglou replied:
`isAutomatticTeamMember` is true / false depending on this wpcom function `get_current_user_teams` which actually uses `is_automattician( $this->current_user->ID )` condition.

cpapazoglou replied:
Using the Proxied condition in all environments (Calypso, Simple, Atomic) was an attempt that had only one fault. Atomic doesn't have a `is_supports_session()` check. We had to ditch that so that HEs on support session experience the customers variant. In case you have an idea that we have missed, please let's discuss it!

tyxla reviewed (dismissed):
This is testing well for me 👍  Thanks!

I've asked a couple of questions that are worth discussing before shipping this, but I don't consider them blockers for merging.

getdave commented:
I'd feel most comfortable with a review from @cpapazoglou before this is merged as he's been the most involved in the "gating" of this feature.

@cpapazoglou feel free to remove the `DO NOT MERGE` label once you're done.

cpapazoglou commented:
Please let's take the decision here https://github.com/Automattic/wp-calypso/issues/49123#issuecomment-765353474.

cpapazoglou commented on client/layout/index.jsx:
```diff
@@ -178,13 +179,6 @@ class Layout extends Component {
 		};
 
 		const { shouldShowAppBanner } = this.props;
-
-		if ( isAutomatticTeamMember( this.props.teams ) ) {
-			config.enable( 'nav-unification' );
-		} else {
-			config.disable( 'nav-unification' );
-		}
-
```
We cannot rely on the feature-flag. We want this feature to be enabled also in production for a12s. This is very useful for testing from mobile devices or also in every-day use which in turn increases the chances to find bugs.

tyxla replied:
As invasive as it feels at first glance, this makes sense for such a huge-impact change. Thanks for clarifying, that sounds good to me 👍

Commit:
Properly allows the feature to be disabled

cpapazoglou commented:
I have made the necessary adjustments in case we decide to proceed with this 3285855

cpapazoglou commented:
> Even given we have the special instructions to disable it for customers in `<Layout />`, I don't feel it's safe to leave it to `true` for `production` under any circumstances. Do you have any alternatives to suggest?

Me neither! An other approach would require a check`'production' !== process.env.NODE_ENV` and it will allow you to disable the feature in all environments except production. In a quick search, I found a lot of `process.env.NODE_ENV` checks would it be better to export the current environment (eg `isProduction()`) in 
https://github.com/Automattic/wp-calypso/blob/cf8f20997de8ab758a9050457d68bc083d5c69d4/client/config/index.js#L1-L77 ?

I cannot decide, whether adding such a function would be more performant ( less KB in codebase ) than using directly `process.env.NODE_ENV`...

mmtr commented:
> Even given we have the special instructions to disable it for customers in `<Layout />`, I don't feel it's safe to leave it to `true` for `production` under any circumstances. Do you have any alternatives to suggest?

I agree. Most of a12s do not use production actually, since we're proxied most of the time and thus end up reaching the staging environment.

In my opinion, enabling the feature on staging is enough. 

EDIT: Just read the reasoning about this on https://github.com/Automattic/wp-calypso/pull/49136#discussion_r562595352. I understand it, but I don't think we should enable the feature on production. The risk of accidentally expose it to final customers becomes higher.

Commit:
Disable the feature in production by default

cpapazoglou commented:
I pushed 8a57011 that:
- has the feature disabled in Production by default but enables it only for a12s.
- allows a12s to disable the feature in all environments except production.

mmtr commented on client/layout/index.jsx:
```diff
@@ -147,6 +147,23 @@ class Layout extends Component {
+		// This is temporary helper function until we have rolled out to 100% of customers.
+		const isNavUnificationEnabled = () => {
+			// Having the feature enabled by default in all environments, will let anyone use ?flags=-nav-unification to temporary disable it.
+			if ( 'production' !== process.env.NODE_ENV && ! config.isEnabled( 'nav-unification' ) ) {
```
Why is the `'production' !== process.env.NODE_ENV` check needed? Would not that prevent a12s from disabling the nav unification when `NODE_ENV` is set to `production`?

cpapazoglou replied:
Yes indeed. Because of the fact that we now set `"nav-unification": false` in production we need an escape mechanism so that it is still enabled for all a12s who don't use `?flags=-nav-unification`. Again, the `isAutomatticTeamMember` is the only consistent way of having the feature enabled across all infrastructures ( Calypso, Simple, Atomic) so enabling the feature depending on this check on all Calypso environments should be the best call.

mmtr replied:
Do we know if `NODE_ENV` is set to `production` for a12s visiting wordpress.com while proxied? That's technically the staging environment, so I wonder if in these cases `NODE_ENV` has a different value like `staging` or `stage`.

I have the feeling that the current feature flags system is not suited for our needs. Why don't we completely avoid it and use a customized one which is based on [`sessionStorage`](https://developer.mozilla.org/en-US/docs/Web/API/Window/sessionStorage)? 

```js
const isNavUnificationEnabled = () => {
  const STORAGE_KEY = 'nav_unification_enabled';

  // Explicitly disabled by a `?disable-nav-unification` query param.
  if ( ( new URL( document.location ) ).searchParams.has( 'disable-nav-unification' ) ) {
    window.sessionStorage.setItem( STORAGE_KEY, false );
    return false;
  }
  
  // Check if has been already enabled/disabled on current session.
  const isEnabled = window.sessionStorage.getItem( STORAGE_KEY );
  if ( isEnabled !== null ) {
    return isEnabled;
  }

  // Enable for all a12s.
  if ( this.props.teams.length && isAutomatticTeamMember( this.props.teams ) ) {
    window.sessionStorage.setItem( STORAGE_KEY, true );
    return true;
  }
  
  return false;
};
```

cpapazoglou replied:
`NODE_ENV` is set to `staging` for a12s visiting wordpress.com while proxied and that's why the will get the chance to disable the feature 

```
if ( 'production' !== process.env.NODE_ENV && ! config.isEnabled( 'nav-unification' ) ) {
  return false;
}
```

Why doesn't the flag system help us here? Instead of using it per environment, we are using it internally (with a custom condition) so that we avoid bugs that a "new system" may include. Also, if we decide changing it we would also need to change all `nav-unification` occurrences which depend on it. I am not that such a temporary solution deserves that effort. Coding it seems super simple but testing it is certainly not.

mmtr replied:
> NODE_ENV is set to staging for a12s visiting wordpress.com

Cool, thanks for confirming. I don't know why I was under the impression that we were using `production` as well on staging. The `'production' !== process.env.NODE_ENV && ! config.isEnabled( 'nav-unification' )` check makes sense now 👍

tyxla commented on client/layout/index.jsx:
```diff
@@ -147,6 +147,23 @@ class Layout extends Component {
 
 	render() {
+		// This is temporary helper function until we have rolled out to 100% of customers.
+		const isNavUnificationEnabled = () => {
```
Is this the right place for this logic? Several things hint that it's not, but let's talk it through and decide. I have some thoughts and would love to hear your thoughts about it.

1. It seems like `isAutomatticTeamMember` relies on the `QueryReaderTeams` triggering a fetch, then receiving that data. This means that if we open `wordpress.com` as a logged-in user, but with a clean Redux store (try it out yourself), we'll see nav unification always disabled. Is that the expected behavior? This also has several sub-questions:
* If we move the check earlier, won't we have the same problem because the teams' data will still not be fetched?
* If we move it later, won't it cause a weird experience when it loads as non-nav-unification and changes the nav to the nav unification one later?
* In any case, this behavior should be consistent and should yield the same results, regardless of whether we're using it with a clean Redux store or not.
2. This currently happens inside the `<Layout />` component, and I'm not sure this is the right place for it. I realize it controls visual changes, so having it in the layout component makes sense, but at the same time, this changes a `config` var. Anything that was called prior to the rendering of `Layout` will not use the proper value, and that might cause unexpected behavior.
3. There are multiple locations where we call just `isEnabled( 'nav-unification' )` and rely on it. Can we somehow guarantee all of those instances are called **after** this logic? Otherwise, we're poised to receiving inconsistent results.
4. Currently, logged out experience displays the old navigation. At the same time, after we login, we see the new one. Have we considered the implications of this difference in navigations after we've merged this for all users? Will all users have to switch from one nav to another? This feels like a regression to me.
5. Minor: this function is declared inside a method and will be redefined on every re-render. Do we want this, or should we declare it inside its own method in the component?

mmtr replied:
> 1. [...] we'll see nav unification always disabled. Is that the expected behavior?

I think this it ok for now. We are not interested on making sure that all a12s get a consistent nav experience always, we are only interested on getting the most feedback we can about the new nav experience. Of course we're aiming for enabling the new nav experience every time that it's possible, but we can live with some edge cases in where the old nav experience will be incorrectly shown (as long as they are edge cases and very unlikely to happen).

> 2. having it in the layout component makes sense, but at the same time, this changes a `config` var
> 3. There are multiple locations where we call just `isEnabled( 'nav-unification' )` and rely on it

~[I also have concerns about dynamically changing the `config` vars](https://github.com/Automattic/wp-calypso/pull/49136#discussion_r563780120). I would consider revisiting the approach and implement something that does not rely on `config` vars since that will simplify things a lot. @cpapazoglou I'm happy to push an update looking into this direction if you don't mind.~ Clarified in https://github.com/Automattic/wp-calypso/pull/49136#discussion_r564505110.

> 4. Currently, logged out experience displays the old navigation

Probably we're missing a scenario here. How a logged out user can see any navigation at all? We thought that the navigation (either new or old) is only displayed to logged in users.

> 5. this function is declared inside a method and will be redefined on every re-render.

Good catch. We'll make sure it's declared outside.

tyxla replied:
In that case, with this in mind and https://github.com/Automattic/wp-calypso/pull/49136#discussion_r564505110 I'm fine with this as a temporary solution to make testing for a12s easier, more reliable while still keeping it possible to opt-out. 👍

cpapazoglou replied:
> Minor: this function is declared inside a method and will be redefined on every re-render. Do we want this, or should we declare it inside its own method in the component?

I have moved the function outside of `render`.

> Currently, logged out experience displays the old navigation. At the same time, after we login, we see the new one. Have we considered the implications of this difference in navigations after we've merged this for all users? Will all users have to switch from one nav to another? This feels like a regression to me.

This is inevitable till we roll out to 100% and thus remove any conditionals. Currently we need a user to determine whether the feature is enabled or not.

> There are multiple locations where we call just isEnabled( 'nav-unification' ) and rely on it. Can we somehow guarantee all of those instances are called after this logic? Otherwise, we're poised to receiving inconsistent results.

All components using `isEnabled( 'nav-unification' )` are currently descendants of `layout`

> It seems like isAutomatticTeamMember relies on the QueryReaderTeams triggering a fetch, then receiving that data. This means that if we open wordpress.com as a logged-in user, but with a clean Redux store (try it out yourself), we'll see nav unification always disabled. Is that the expected behavior? 

This should be fixed now. I was disabling the feature even when `this.props.teams.length === 0` resulting in bailing out early. Make sure you stop your process and restart by `yarn build && yarn start` as it heavily depends on the `development.json` which now defaults to true.

Commit:
Moves function outside render, returns early when no teams.

cpapazoglou commented:
We have a problem here... Enabling the feature for all environments results in e2e tests failures... Adjusting e2e tests to nav-unification is a pending issue https://github.com/Automattic/wp-calypso/issues/49250. cc @getdave 

In order that skip this problem, I had to disable the feature from all environments and only enable it depending on the user. Which is the way we will also handle tests pciE2j-bk-p2#comment-264.

The escape mechanism will be `?disable-nav-unification` in the querystring.

Commit:
Disables feature for all environments and when ?disable-nav-unification exists.

mmtr commented on client/layout/index.jsx:
```diff
@@ -146,7 +146,31 @@ class Layout extends Component {
+			return;
+		}
+
+		// Having the feature enabled by default in all environments, will let anyone use ?flags=-nav-unification to temporary disable it.
```
```suggestion
		// Having the feature enabled by default in all environments, will let anyone use ?disable-nav-unification to temporary disable it.
```

mmtr commented on client/layout/index.jsx:
```diff
@@ -146,7 +146,31 @@ class Layout extends Component {
+		// Having the feature enabled by default in all environments, will let anyone use ?flags=-nav-unification to temporary disable it.
+		// We still have the feature disabled in production as safety mechanism for all customers.
+		if ( new URL( document.location ).searchParams.has( 'disable-nav-unification' ) ) {
+			return;
```
```suggestion
			return config.disable( 'nav-unification' );
```

Otherwise, the nav unification is enabled after navigating to a different page:

![Jan-26-2021 18-06-12](https://user-images.githubusercontent.com/1233880/105878346-3aa41e00-6001-11eb-981c-22b03e13e39a.gif)

mmtr replied:
Actually, not sure if my suggested change will fix this, since the `isNavUnificationEnabled` function will be re-evaluated in the next render and the query param won't be present. Maybe we can fix by moving the `this. isNavUnificationEnabled()` from `render` to `componentDidMount`?

cpapazoglou replied:
Nope, nesting it in `componentDidMount` doesn't work consistently cause `teams` is updated after the layout mounts. What I did is:
- use `UNSAFE_componentWillMount` for initial load (eg `teams` is already in indexed DB)
- use `componentDidUpdate` for catching `teams` update and only call `this.isNavUnificationEnabled();` if the `teams` object has changed.

7c46153

mmtr reviewed (commented):
Pretty close, @cpapazoglou! Thanks for noticing the e2e tests breakage and working on a workaround.

This worked quite well for me, except that the feature does not remain disabled after navigating to a different page. Suggested a possible solution for this.

cpapazoglou committed:
Updates comment

Co-authored-by: Miguel Torres <miguel.torres@automattic.com>

Commit:
- use UNSAFE_componentWillMount for initial load (eg teams is already in indexed DB)
- use componentDidUpdate for catching teams update and only call this.isNavUnificationEnabled(); if the teams object has changed.

mmtr reviewed (approved):
Nice work @cpapazoglou, this tests pretty well for me. Thanks for working so hard on this ❤️
